StableMIR: Implement `CompilerInterface`

makai410 opened the pull request:
This PR implements part of [the document](https://hackmd.io/@celinaval/H1lJBGse0).

With `TablesWrapper` wrapped by `CompilerInterface`, the stable-mir's TLV stores a pointer to `CompilerInterface`, while the rustc-specific TLV stores a pointer to tables.

makai410 commented:
r? @celinval btw should we create a tracking issue for this?

makai410 commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -233,7 +234,12 @@ where
     }));
-    stable_mir::compiler_interface::run(&tables, || init(&tables, f))
+
+    let interface = CompilerInterface { cx: tables };
```
I put the interface here since I'd prefer not to refactor the launch process now. It would be more appropriate to do this at the end.

celinval commented on compiler/rustc_smir/src/stable_mir/compiler_interface.rs:
```diff
@@ -258,31 +259,464 @@ pub trait Context {
-// A thread local variable that stores a pointer to the tables mapping between TyCtxt
-// datastructures and stable MIR datastructures
+/// Stable-MIR’s interface to compiler internals.
+/// All internal queries must go through [`Context`].
```
Can we remove `Context` in this PR?

Also, I think it may be worth documenting this a bit better, so people know why this is here. Something on the lines that this structure will serve as a bridge between StableMIR and the rustc_smir which will provide similar APIs but based on internal rustc constructs.

makai410 replied:
> Can we remove `Context` in this PR?

Sorry that i didn't get it. You mean we directly write something like impl<'tcx> TablesWrapper<'tcx>, then move everything that used to be in Context to there?

makai410 replied:
Oh and then we create a brand-new `Context` in `rustc_smir`, gradually move methods from `TablesWrapper` to it until `TablesWrapper` is completely hollowed out.

celinval replied:
Also, feel free to improve the name scheme, it's not my strongest skill. LoL

makai410 replied:
hmmm what if we change `CompilerInterface` to `SmirInterface`?🤔

celinval commented on compiler/rustc_smir/src/stable_mir/compiler_interface.rs:
```diff
@@ -258,31 +259,464 @@ pub trait Context {
 scoped_tls::scoped_thread_local!(static TLV: Cell<*const ()>);
 
-pub fn run<F, T>(context: &dyn Context, f: F) -> Result<T, Error>
+pub fn run<'tcx, T, F>(interface: &CompilerInterface<'tcx>, f: F) -> Result<T, Error>
```
@oli-obk Any chance you can take a look at this TLV magic? I could use a second pair of eyes here. Thanks!

celinval reviewed (commented):
Awesome!

makai410 commented on compiler/rustc_smir/src/stable_mir/compiler_interface.rs:
```diff
@@ -258,31 +259,464 @@ pub trait Context {
+/// Stable-MIR’s interface to compiler internals.
+/// All internal queries must go through [`Context`].
+///
+/// Do not use this directly.
```
hmmm tbh it is not. We directly create a `CompilerInterface` in the `rustc_internal::run()` function for now, which is also called by `pretty::write_smir_pretty()`, not only the `run_driver!`.

makai410 replied:
but yeah, maybe we can make it into a macro

makai410 replied:
Oh my bad! I get it wrong...

celinval commented on compiler/rustc_smir/src/rustc_smir/context.rs:
```diff
@@ -37,8 +36,11 @@ use crate::rustc_smir::builder::BodyBuilder;
+pub(crate) struct Context<'tcx>(pub RefCell<Tables<'tcx>>);
+
+impl<'tcx> Context<'tcx> {
+    pub(crate) fn target_info(&self) -> MachineInfo {
```
You might as well make these public so we don't have to update the visibility when we pull the StableMir stuff back to its crate

makai410 replied:
done

makai410 committed:
Implement `SmirInterface`
 - With `Context` wrapped by `SmirInterface`, the stable-mir's TLV stores a pointer to `SmirInterface`, while the rustc-specific TLV stores a pointer to tables.
 - This PR make the `rustc_smir` mod public.

celinval commented on compiler/rustc_smir/src/stable_mir/compiler_interface.rs:
```diff
@@ -22,47 +23,116 @@ use stable_mir::{
+/// similar APIs but based on internal rustc constructs.
+///
+/// Do not use this directly. This is currently used in the macro expansion.
+pub struct SmirInterface<'tcx> {
```
On the other hand, this one should be `pub(crate)`, since users of `stable_mir` should not interact with it.

celinval commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -18,9 +18,10 @@ use rustc_span::def_id::{CrateNum, DefId};
 use stable_mir::ty::IndexedVal;
 
-use crate::rustc_smir::context::TablesWrapper;
+use crate::rustc_smir::context::Context;
```
I wonder if we should call this `SmirContext`

makai410 replied:
SmirCtxt?

celinval reviewed (approved):
Looks good! Thanks.

makai410 committed:
Make `SmirInterface` pub(crate) and rename `Context` to `SmirContext`

Co-authored-by: Celina G. Val <celinval@amazon.com>

celinval commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 5b390cdc58fb7f0f0a162f96c55b36641666cbb7 has been approved by `celinval`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=celinval 5b390cdc58fb7f0f0a162f96c55b36641666cbb7 -->
<!-- homu: {"type":"Approved","sha":"5b390cdc58fb7f0f0a162f96c55b36641666cbb7","approver":"celinval","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
